### PR TITLE
Create an injectable factory for application packages.

### PR DESCRIPTION
**Pull request by skycomposter**

### CODE REVIEW

**Review by jonahwilliams — APPROVED** [review]
LGTM

If you sync to the latest master that should include a skip for expression_evaluation_test.dart which is failing in your PR

**Comment by mehmetf** [comment]
@jonahwilliams out of curiosity, you message seems to imply that all checks are run at the commit of the *client* and not at head?

Please also provide @skycomposter with instructions on how to sync to head. I usually do this with an clean branch but not sure it would work if I already have a PR out:

```
git fetch upstream
git rebase upstream/master
git push -f origin master
```

**Comment by jonahwilliams** [comment]
Yes, I believe the presubmit checks aren't merged. At one point we had merge PR checks (cc @gspencergoog).

I wouldn't worry about rebasing, assuming you branched off of `upstream/master`:

```
git fetch upstream
git pull upstream master
git push origin application-package
```

File: packages/flutter_tools/lib/src/globals.dart [hunk 1]
```diff
@@ -13,6 +14,7 @@ Logger get logger => context[Logger];
 Cache get cache => Cache.instance;
 Config get config => Config.instance;
 Artifacts get artifacts => Artifacts.instance;
+ApplicationPackageFactory get applicationPackageFactory => ApplicationPackageFactory.instance;
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/globals.dart]: Let's just refer to `ApplicationPackageFactory.instance` explicitly where we need it - we're trying to reduce the amount of globals we have.

File: packages/flutter_tools/lib/src/application_package.dart [hunk 2]
```diff
@@ -20,6 +21,37 @@ import 'ios/plist_utils.dart' as plist;
 import 'project.dart';
 import 'tester/flutter_tester.dart';
 
+class ApplicationPackageFactory {
+  static ApplicationPackageFactory get instance => context[ApplicationPackageFactory];
+
+  Future<ApplicationPackage> getApplicationPackageForPlatform(
```
Comment (tvolkert) [inline packages/flutter_tools/lib/src/application_package.dart]: This method name seems a little overly verbose now that it's in a `ApplicationPackageFactory` class.  Perhaps just `getPackageForPlatform()`?

**Review by tvolkert — APPROVED** [review]

**Pull request merged**